minimal_data_sharing/minimal_lock_free_queue.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 Carlos San Vicente
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+
+#include "rclcpp/rclcpp.hpp"
+#include "trajectory_msgs/msg/joint_trajectory.hpp"
+
+#include <iceoryx_utils/concurrent/lockfree_queue.hpp>
+
+
+using namespace std::chrono_literals;
+
+class TrajectoryGenerator : public rclcpp::Node
+{
+public:
+  TrajectoryGenerator()
+  : Node("trajectory_generator"), count_(0)
+  {
+    constexpr std::size_t TRAJECTORY_SIZE = 10;
+    publisher_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>("trajectory", 10);
+    auto timer_callback =
+      [this]() -> void {
+        auto message = trajectory_msgs::msg::JointTrajectory();
+        message.points.reserve(TRAJECTORY_SIZE);
+        for (std::size_t i = 0; i < TRAJECTORY_SIZE; i++) {
+          trajectory_msgs::msg::JointTrajectoryPoint point;
+          point.positions.push_back(count_++);
+          message.points.push_back(point);
+        }
+        RCLCPP_INFO(
+          this->get_logger(), "Publish trajectory with %zu points",
+          message.points.size());
+        this->publisher_->publish(message);
+      };
+    timer_ = this->create_wall_timer(1s, timer_callback);
+  }
+
+private:
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr publisher_;
+  size_t count_;
+};
+
+class TrajectoryController : public rclcpp::Node
+{
+public:
+  using JointTrajectoryPointQueue =
+    iox::concurrent::LockFreeQueue<trajectory_msgs::msg::JointTrajectoryPoint, 100>;
```

Comment:
Might want to quickly document why you chose 100 to avoid copy pasting from the examples without thinking. In this case, I see the publisher is executing at 1Hz, and the subscriber is executing at 10Hz, so theoretically we can never fill up 100 unless something bad has happened and the subscriber failed to execute for 100 seconds. (in fact, you might want to lower the capacity to make error detection faster?)

```suggestion
  // Chose a capacity of 100 as the rate of push is 1Hz and the rate of pop is 10Hz. 
  // Theoretically, there should never be a buffer overrun unless the pop operation froze for 
  // 100 seconds.
  using JointTrajectoryPointQueue =
    iox::concurrent::LockFreeQueue<trajectory_msgs::msg::JointTrajectoryPoint, 100>;
```

Also, is there a reason for this type alias to be public? I don't see it used outside of this class. It could just be defined near the queue definition itself so the code is closer together and easier to read?

Reply:
>Might want to quickly document why you chose 100 to avoid copy pasting from the examples without thinking. In this case, I see the publisher is executing at 1Hz, and the subscriber is executing at 10Hz, so theoretically we can never fill up 100 unless something bad has happened and the subscriber failed to execute for 100 seconds. (in fact, you might want to lower the capacity to make error detection faster?)

My first idea was that the TrajectoryController receives `JointTrajectory` messages at arbitrary times with arbitrary (but bounded) JointTrajectoryPoint`s. The use case would be, some GUI or higher-level task manager sending trajectories depending on different events. For example, the user sends a trajectory to pick an object, we run a program to execute a sequence of  trajectories or event better, a a real-time planner sending trajectories based on computer vision. Using a fixed publishing period was just a simple way to run the example. 

I guess we can make this more explcit by simulating one of this use cases so we don't use a fixed publishing period. Then we can reason about the queue size, and why we decided to buffer this amount. 

>Also, is there a reason for this type alias to be public? I don't see it used outside of this class. It could just be defined near the queue definition itself so the code is closer together and easier to read?

Yes, I think it makes sense to move it next to the private section if it's not used by any API.